Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1289 @@
+	pItem->pUIIcon->SetMoveRect(pArea->GetRegion());
+}
+
+bool CUIItemUpgrade::HandleUpgradeAreaDrop(__IconItemSkill* spItem)
+{
+	// Only Upgrade and Unique items can be dropped here
+	if (!IsAllowedUpgradeItem(spItem))
+		return false;
+
+	// Move the item to the upgrade slot
+	m_pUpgradeItemSlot = spItem;
+
+	// Remove the item from inventory
+	for (int i = 0; i < MAX_ITEM_INVENTORY; ++i)
+	{
+		if (m_pMyUpgradeInv[i] == spItem)
+		{
+			m_pMyUpgradeInv[i] = nullptr;
+			m_iUpgradeSlotInvPos[0] = i; // m_pUpgradeItemSlot position
+			break;
+		}
+	}
+
+	// Update the item's UI position
+	if (m_pAreaUpgrade)
+	{
+		spItem->pUIIcon->SetRegion(m_pAreaUpgrade->GetRegion());
+		spItem->pUIIcon->SetMoveRect(m_pAreaUpgrade->GetRegion());
+		spItem->pUIIcon->SetParent(this);
+	}
+
+	return true;
+}
+
+bool CUIItemUpgrade::IsSlotCompatible(__IconItemSkill* pSrc, int iDestiOrder)
+{
+	if (!IsUpgradeScroll(pSrc->pItemBasic->dwEffectID2))
+		return false;
+
+	bool m_bhasTrina = false;
+	bool m_bhasScroll = false;
+
+	for (int k = 0; k < MAX_ITEM_UPGRADE_SLOT; ++k)
+	{
+		if (m_pMyUpgradeSLot[k])
+		{
+			if (IsTrina(m_pMyUpgradeSLot[k]->pItemBasic->dwID))
+			{
+				m_bhasTrina = true;
+			}
+			else 
+				m_bhasScroll = true;
+		}
+	}
+	if (m_bhasTrina && IsTrina(pSrc->pItemBasic->dwID))
+		return false;
+	if (m_bhasScroll && !IsTrina(pSrc->pItemBasic->dwID))
+		return false;
+	if (m_bhasTrina && m_bhasScroll)
+		return false;
+	return true;
+}
+
+bool CUIItemUpgrade::HandleSlotDrop(__IconItemSkill* spItem, int iDestiOrder)
+{
+	if (iDestiOrder == -1 || m_pMyUpgradeSLot[iDestiOrder] != nullptr)
+		return false;
+
+	int iSourceOrder = GetItemiOrder(spItem, UIWND_DISTRICT_UPGRADE_INV);
+	if (iSourceOrder == -1)
+		return false;
+
+	__IconItemSkill* pSrc = m_pMyUpgradeInv[iSourceOrder];
+	if (!IsSlotCompatible(pSrc, iDestiOrder))
+		return false;
+
+	// Handle countable items
+	if (pSrc->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE ||
+		pSrc->pItemBasic->byContable == UIITEM_TYPE_COUNTABLE_SMALL)
+	{
+		if (pSrc->iCount > 1)
+		{
+			// Create a new icon, put it in 1 slot, reduce the number in the inventory
+			__IconItemSkill* pNew = CreateIconFromSource(pSrc, 1);
+			char szID[32];
+			sprintf(szID, "a_m_%d", iDestiOrder);
+			CN3UIArea* pSlotArea = (CN3UIArea*)GetChildByID(szID);
+			SetupIconArea(pNew, pSlotArea);
+			m_pMyUpgradeSLot[iDestiOrder] = pNew;
+			pSrc->iCount -= 1;
+		}
+		else
+		{
+			// If the last one, move directly
+			m_pMyUpgradeSLot[iDestiOrder] = pSrc;
+			m_pMyUpgradeInv[iSourceOrder] = nullptr;
+			char szID[32];
+			sprintf(szID, "a_m_%d", iDestiOrder);
+			CN3UIArea* pSlotArea = (CN3UIArea*)GetChildByID(szID);
+			SetupIconArea(pSrc, pSlotArea);
+		}
+	}
+	else
+	{
+		// If is not countable item, just move it
+		m_pMyUpgradeSLot[iDestiOrder] = pSrc;
+		m_pMyUpgradeInv[iSourceOrder] = nullptr;
+		char szID[32];
+		sprintf(szID, "a_m_%d", iDestiOrder);
+		CN3UIArea* pSlotArea = (CN3UIArea*)GetChildByID(szID);
```

Comment:
Don't lookup children on-demand. Pre-load them in `Load()` (ideally) or `InitIconWnd()`.

Note that all of these lookups and `SetupIconArea()` calls here could've been moved to handle after these branches. They're the only reason `SetupIconArea()` even exists, and you could've just handled this logic once anyway.